rfcs/0000-structured-logging.md

```diff
@@ -0,0 +1,2189 @@
+# Summary
+[summary]: #summary
+
+Add support for structured logging to the `log` crate in both `std` and `no_std` environments, allowing log records to carry typed data beyond a textual message. This document serves as an introduction to what structured logging is all about, and as an RFC for an implementation in the `log` crate.
+
+`log` will provide a lightweight fundamental serialization API out-of-the-box that can be expanded to full `serde` support using crate features. It doesn't require crates implement any traits from `log` before their datastructures can be used in the `log!` macros. Instead, crates should implement `std::fmt::Debug` and `serde::Serialize`, which are traits they should by considering already. The API is heavily inspired by the `slog` logging framework. See [the implications for `log` users](#implications-for-dependents) for some more details.
```

Comment:
> This paragraph conflicts with itself, it says serde is optional follow by "crates should implement std::fmt::Debug and serde::Serialize".

Yeh the definition of _optional_ here might be a little unclear :) By optional I mean anyone can depend on `log` and use the `log!` macros to log structured data, or implement `Log` and work with the structured data on a record without `serde` entering the equation at all if you don't opt in to the `kv_serde` Cargo feature. The `kv_serde` feature adds `serde` support, but is totally optional. Any crate calling `log!` macros can work without this feature. Any crate implementing `Log` and working with the key-value pairs can work without this feature. Any logs produced when `serde` support is opted into can be visited by any `Log` implementation, even if they haven't opted into the feature.

The limitation of structured logging in `log` without the `kv_serde` feature is that only a set of common types from the standard library can be logged as structured key-value pairs. Most types in the ecosystem like `Uuid`, `Url`, and `DateTime<Utc>` implement `serde::Serialize`, so this RFC proposes leveraging that existing investment in serialization boilerplate for libraries that exist today, as well as libraries that are getting built in the future, instead of inventing a new framework that they will need to support _in addition to_ `serde`.

Reply:
Your comment makes a lot of sense, maybe its worth it to repeat that in the RFC.